File: src/Asm.MooBank.Web/MooBankApp/src/models/Transaction.ts
import { v4 as guid } from "uuid";

import { Tag } from "./Tag";
import { emptyGuid } from "@andrewmclachlan/mooapp";

enum TransactionTypesEnum {
    Credit = 1,
    Debit = 2,
    RecurringCredit = 3,
    RecurringDebit = 4,
    BalanceAdjustment = 5,
}

export const TransactionTypes = ["Credit", "Debit", "RecurringCredit", "RecurringDebit", "BalanceAdjustment"] as TransactionType[];
export type TransactionType = "Credit" | "Debit" | "RecurringCredit" | "RecurringDebit" | "BalanceAdjustment";


export const isCredit = (transactionType: TransactionType) => TransactionTypesEnum[transactionType] % 2 === 1;

export const isDebit = (transactionType: TransactionType) => TransactionTypesEnum[transactionType] % 2 === 0;

export interface Transaction {
    id: string;
    accountId: string;
    amount: number;
    netAmount: number;
    description: string;
    location?: string;
    reference?: string;
    accountHolderName?: string;
    purchaseDate?: string;
    transactionTime: string;
    transactionType: TransactionType;
    tags: Tag[];
    notes?: string;
    splits: TransactionSplit[];
    offsetFor: TransactionOffset[];
    extraInfo: any;
}

export interface TransactionUpdate {
    notes?: string;
    splits?: TransactionSplit[];
}

export interface TransactionOffsetUpdate {
    transactionOffsetId: string;
    amount: number;
}

export interface TransactionOffset {
    transaction: Transaction;
    amount: number;
}

export interface TransactionSplit {
    id: string;
    amount: number;
    offsetBy: TransactionOffset[];
    tags: Tag[];
}

export const emptyTransactionSplit: TransactionSplit = {
    id: emptyGuid,
    amount: 0,
    offsetBy: [],
    tags: [],
};

export const getSplitTotal = (splits: TransactionSplit[]) => splits.reduce((total, split) => total + split.amount, 0);

File: src/Asm.MooBank.Modules.Transactions/Models/Transaction.cs
using Asm.MooBank.Models;

namespace Asm.MooBank.Modules.Transactions.Models;

public partial record Transaction
{
    public Guid Id { get; set; } = new Guid();
    public Guid AccountId { get; set; }
    public decimal Amount { get; set; }

    public string? Description { get; set; }

    public string? Location { get; set; }

    public string? AccountHolderName { get; set; }

    public string? Reference { get; set; }
    public string? Notes { get; set; }

    public bool ExcludeFromReporting { get; set; }
    public DateTime TransactionTime { get; set; }

    public DateTime? PurchaseDate { get; set; }

    public TransactionType TransactionType { get; set; }

    public IEnumerable<Tag> Tags { get; set; } = Enumerable.Empty<Tag>();

    public IEnumerable<TransactionSplit> Splits { get; set; } = Enumerable.Empty<TransactionSplit>();

    public IEnumerable<TransactionOffsetFor> OffsetFor { get; set; } = Enumerable.Empty<TransactionOffsetFor>();

    public object? ExtraInfo { get; set; }
}
